### PR TITLE
20200925 latest

### PR DESCRIPTION
**Pull request by eoliveros**
i thought i created this yesterday but may have not. so creating this PR.

### CODE REVIEW

File: app_core.py [hunk 1]
```diff
@@ -49,6 +49,8 @@
     app.config["SENDER_BANK_ACCOUNT"] = os.getenv("SENDER_BANK_ACCOUNT")
 if os.getenv("SENDER_NAME"):
     app.config["SENDER_NAME"] = os.getenv("SENDER_NAME")
+if os.getenv("HTTPS_ADDRESS"):
```
Comment (djpnewton) [inline app_core.py]: Can we change this to 'SITE_URL'

File: config.py [hunk 2]
```diff
@@ -21,12 +21,16 @@
 SECURITY_POST_REGISTER_VIEW = "/admin/"
 
 # Flask-Security features
-SECURITY_CONFIRMABLE = True
```
Comment (djpnewton) [inline config.py:24]: Can we change this to be true if an environment variable is present (eg 'ALLOW_USER_REGISTRATION')

File: config.py [hunk 3]
```diff
@@ -21,12 +21,16 @@
...
+SECURITY_REGISTERABLE = False
 SECURITY_RECOVERABLE = True
 SECURITY_CHANGEABLE = True
 SQLALCHEMY_TRACK_MODIFICATIONS = False
 
 # Flask-Mail config
 MAIL_SENDGRID_API_KEY = "XXXX"
 SECURITY_EMAIL_SENDER = "noreply@retail.zap.me"
+
+SETTLEMENT_ADDRESS = ""
```
Comment (djpnewton) [inline config.py]: why are these added here?

File: models.py [hunk 4]
```diff
@@ -69,6 +72,12 @@ def __init__(self, **kwargs):
         self.merchant_code = generate_key(4)
         super().__init__(**kwargs)
 
+    def generate_defaults(self):
```
Comment (djpnewton) [inline models.py]: change this function to 'on_admin_created', remove merchant_code initialization because it is already done in the constructor

File: templates/admin/index.html [hunk 5]
```diff
@@ -6,7 +6,7 @@
         <div class="col-sm-10 col-sm-offset-1">
             {% if not current_user.is_authenticated %}
             <p>
-                <a class="btn btn-primary" href="{{ url_for('security.login') }}">login</a> <a class="btn btn-default" href="{{ url_for('security.register') }}">register</a>
```
Comment (djpnewton) [inline templates/admin/index.html:9]: change this to only be shown if `app.config["SECURITY_REGISTERABLE"] == True`

File: templates/security/register_user.html [hunk 6]
```diff
@@ -20,4 +20,5 @@ <h1>Register</h1>
         </div>
     </div>
 </div>
-{% endblock body %}
```
Comment (djpnewton) [inline templates/security/register_user.html:23]: ?

File: utils.py [hunk 7]
```diff
@@ -81,3 +84,11 @@ def blockchain_transactions(logger, node, wallet_address, limit, after=None):
 def apply_merchant_rate(amount, rate):
     amount_nzd = (amount / float(1 + rate))
     return amount_nzd
+
+def is_email(email):
+    return re.match("[^@]+@[^@]+\.[^@]+", email)
+
+def generate_random_password(length):
```
Comment (djpnewton) [inline utils.py:91]: change this to use a crypto secure source of randomness: https://docs.python.org/3/library/secrets.html#recipes-and-best-practices

**Review by djpnewton — CHANGES_REQUESTED** [review]

File: models.py (hunk 4 above)
Comment (eoliveros) [reply models.py]: im pretty sure the merchant_code doesn't get generated when going via create form.

Jsut tested this by commenting the line:
```
    def generate_defaults(self):
        #self.merchant_code = generate_key(4)
        self.password = encrypt_password(generate_random_password(16))
        self.confirmed_at = datetime.datetime.now()
        self.active = True
```

and the merchant_code is not being triggered. The previous is way must be getting triggered by the registerable? but since its disabled in this instance its not getting triggered. unless something else is generating the merchant_code.


drop the table for the user in my testvm, which basically meant i need to create a new user via CLI.
this is creating the merchant_code automatically. so something is triggering it.

File: config.py (hunk 2 above)
Comment (eoliveros) [reply config.py:24]: Enabling SECURITY_REGISTERABLE must enable SECURITY_CONFIRMABLE as well.

Comment (eoliveros) [reply config.py:24]: added

File: app_core.py (hunk 1 above)
Comment (eoliveros) [reply app_core.py]: changed

File: config.py (hunk 3 above)
Comment (eoliveros) [reply config.py]: because i added it when i was running on my testvm

File: templates/admin/index.html (hunk 5 above)
Comment (eoliveros) [reply templates/admin/index.html:9]: added

File: templates/security/register_user.html (hunk 6 above)
Comment (eoliveros) [reply templates/security/register_user.html:23]: removed the space at the last line.

File: models.py (hunk 4 above)
Comment (eoliveros) [reply models.py]: so it looks like when adding the user via CLI the difference is this:
    if len(sys.argv) > 1:
        if sys.argv[1] == "add_user":
            add_user(sys.argv[2], sys.argv[3])
        if sys.argv[1] == "add_role":
            add_role(sys.argv[2], sys.argv[3])
        if sys.argv[1] == "add_merchant_codes":
            add_merchant_codes()

So the merchant_codes being added when the you're adding the user or role, as long as there is more than 1 argument. Which is different from the when the admin form is adding it.

Comment (djpnewton) [reply models.py]: ok just rename the function then

File: templates/admin/index.html [hunk 8]
```diff
@@ -5,8 +5,12 @@
     <div class="row">
         <div class="col-sm-10 col-sm-offset-1">
             {% if not current_user.is_authenticated %}
-            <p>
+            <p> 
+	        {% if security.registerable %}
```
Comment (djpnewton) [inline templates/admin/index.html]: you can just put the second link in the condition

File: templates/security/register_user.html [hunk 9]
```diff
@@ -21,4 +21,3 @@ <h1>Register</h1>
     </div>
 </div>
 {% endblock body %}
-
```
Comment (djpnewton) [inline templates/security/register_user.html]: extra line break

Comment (eoliveros) [reply templates/security/register_user.html]: i removed the extra line break

File: utils.py (hunk 7 above)
Comment (eoliveros) [reply utils.py:91]: im reading this and from what i could tell is to replace random and use secrets instead.

done.
